pcap/pcap_unix.go

```diff
@@ -343,12 +347,15 @@ func pcapLookupnet(device string) (netp, maskp uint32, err error) {
 
 func (b *BPF) pcapOfflineFilter(ci gopacket.CaptureInfo, data []byte) bool {
 	var hdr C.struct_pcap_pkthdr
+	p := uintptr(unsafe.Pointer(&hdr))
```

Comment:
This can cause some trouble:
According to https://golang.org/pkg/unsafe/ this is illegal to do, because now the compiler is free to think, that you're actually not using `hdr`, after line 354, which can lead to surprising results.

If one reads the documentation closely, one would discover that this is only valid for syscalls (actually functions defined in assembly) then there is a second issue:
Between converting this pointer to `uintptr` and actually accessing it, go might change the stack (stack split etc). But with this construct go doesn't know about the pointer any longer, so it can't adjust it - leading `pcap_offline_filter_escaping` to point to the wrong memory. Actually cgo tries to lock down the stack, but this takes a bit (the cgo call below actually generates new function that contains a pointer check and a call to another autogenerated function that itself calls `cgocall`, which in turn then locks down the stack after some housekeeping) - if we hit this in the small window surprising stuff will happen..
So this can fail in surprising ways.

Solution:
The same thing is also in `pcap_next_ex_escaping` - but there we reference stuff from `*Handle`, which is guaranteed to life on the heap due to the way it is created with e.g. `pcapOpenLive` (there `Handle` escapes to the heap). So still not safe according to the documentation, but at least safe from the things that can happen.

So correct way would be to put the hdr into *BPF, where it is guaranteed to live on the heap.
To keep compatibility with windows this must be done as `pcapPkthdr` which is a typedef to `C.struct_pcap_pkthdr`.

Reply:
Thanks for the explanation and suggestion. I'll handle it.